modal/image.py

```diff
@@ -1222,6 +1222,115 @@ def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
+
+        **Examples**
+
+        Simple installation:
+        ```python
+        image = modal.Image.debian_slim().uv_pip_install("torch==2.7.1", "numpy")
+        ```
+
+        This method assumes that:
+        - Python is on the `$PATH` and dependencies are installed with the first Python on the `$PATH`.
+        - Shell supports backticks for substitution
+        - `which` command is on the `$PATH`
+        """
+
+        def _normalize_items(items, name) -> list[str]:
+            if items is None:
+                return []
+            elif isinstance(items, list):
+                return items
+            else:
+                raise InvalidError(f"{name} must be None or a list of strings")
+
+        pkgs = _flatten_str_args("uv_pip_install", "packages", packages)
+        requirements = _normalize_items(requirements, "requirements")
+
+        if not pkgs and not requirements:
+            return self
+        elif not _validate_packages(pkgs):
+            raise InvalidError(
+                "Package list for `Image.uv_pip_install` cannot contain other arguments;"
+                " try the `extra_options` parameter instead."
+            )
+
+        def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
+            commands = ["FROM base"]
+            UV_ROOT = "/.uv"
+            if uv_version is None:
+                commands.append(f"COPY --from=ghcr.io/astral-sh/uv:latest /uv {UV_ROOT}/uv")
+            else:
+                commands.append(f"COPY --from=ghcr.io/astral-sh/uv:{uv_version} /uv {UV_ROOT}/uv")
+
+            # NOTE: Using `which python` assumes:
+            # - python is on the PATH and uv is installing into the first python in the PATH
+            # - the shell supports backticks for substitution
+            # - `which` command is on the PATH
+            uv_pip_args = ["--python `which python`"]
+            context_files = {}
+
+            if find_links:
+                uv_pip_args.append(f"--find-links {shlex.quote(find_links)}")
+            if index_url:
+                uv_pip_args.append(f"--index-url {shlex.quote(index_url)}")
+            if extra_index_url:
+                uv_pip_args.append(f"--extra-index-url {shlex.quote(extra_index_url)}")
+            if pre:
+                uv_pip_args.append("--prerelease allow")
+            if extra_options:
+                uv_pip_args.append(extra_options)
+
+            if requirements:
+
+                def _generate_paths(req: str) -> dict:
+                    basename = os.path.basename(req)
+                    local_path = os.path.expanduser(req)
+
+                    return {
+                        "basename": basename,
+                        "local_path": local_path,
+                        "context_path": f"./{basename}",
+                        "dest_path": f"{UV_ROOT}/{basename}",
+                    }
+
+                requirement_paths = [_generate_paths(req) for req in requirements]
```

Comment:
Given that `requirements="requirements.txt"` would be an easy mistake and iteration over strings is valid (but would cause a confusing error) should we proactively catch that and fail nicely? (I guess I'd also be more open to `requirements` having a `str | sequence[str]` type since it's not grammatically confused 🤓).

Reply:
oh wait i think that happens in `_normalize_items` actually?

Reply:
Oh I guess that happens in `_normalize_items` actually? Nvm.